keps/sig-node/3542-cri-image-pulling-with-progress-notification/README.md

```diff
@@ -0,0 +1,861 @@
+  - time-based
+  - size-based
+  - none
+- frequency of the updates (if any) respectively to the granularity type:
+  - every N seconds
+  - every N Kibibytes of the total image size being downloaded
+- no-progress timeout
+  - number of seconds during which if completely no data transfer was ongoing, failure should be reported
+
+
+If / when it is possible to reliably determine percentage of the progress in the runtime,
+percentage-based granularity type can be introduced then.
+
+If the client did not specify the preferred notification granularity, default values should be used,
+for instance every Gibibyte downloaded or every 60 seconds of time spent downloading an image.
+
+For kubelet we propose the default granularity to be time-based with 30 seconds interval, so
+Pod object will get an event with image pulling progress every 30 seconds, and setting minimum 
+frequency limits:
+- 10 seconds for the time-baesd frequency
+- 1 Gibibyte for size-based frequency
+
+The default no-progress timeout for kubelet is suggested to be 10 seconds.
+
+The granularity type, interval, and no-progress timeout should be configurable through kubelet-config.
+
+Suggested new Kubelet config fields are these:
+
+       // ImagePullProgressType is the type of ImagePullProgressInterval.
+       // Supported values are
+       // - "time" for tunrime to report progress every ImagePullProgressInterval seconds
+       // - "size" for runtime to report progress every ImagePullProgressInterval <binarySI quantity> (e.g. 1Gi)
+       // - "none" for no progress reporteds
+       // Default: "time"
+       // +optional
+       ImagePullProgressType string `json:"ImagePullProgressType,omitempty"`
+       // ImagePullProgressInterval is a quantity value of how often the runtime should send back to
+       // kubelet image pulling progress reports, that will be published as Pod events.
+       // This option is used together with ImagePullProgressType in ImageService to fill Interval and
+       // GranularityType fields respectively in PullImageWithProgressRequest in CRI PullImageWithProgress.
+       // ImagePullProgressInterval is treated as seconds if ImagePullProgressType is "time".
+       // ImagePullProgressInterval is treated as binarySI quantity if ImagePullProgressType is "size".
+       // call.
+       // Minimum:
+       //   - 10  for "time" ImagePullProgressType
+       //   - 1Gi for "size" ImagePullProgressType
+       // Default: 30
+       // +optional
+       ImagePullProgressInterval string `json:"ImagePullProgressType,omitempty"`
+       // NoProgressTimeout is a number of seconds after which stalled image download should be reported
+       // as an error.
+       // Supported values are:
+       // - 0 for infinity, effectively no timeout
+       // - positive number up to 4294967295 (uint32 max number, approx. 136 years)
+       // Default: 10
+       NoProgressTimeout int32 `json:"NoProgressTimeout,omitempty"`
```

Comment:
consider if this should be optional.. and also used in the pull image request struct

Reply:
Of course.